cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -734,6 +924,19 @@ std::shared_ptr<ScalarFunction> MakeUnarySignedArithmeticFunctionNotNull(
   return func;
 }
 
+// Integer arguments return double values
+template <typename Op>
+std::shared_ptr<ScalarFunction> MakeUnaryIntToDoubleNotNull(std::string name,
+                                                            const FunctionDoc* doc) {
+  auto func = std::make_shared<ArithmeticFunction>(name, Arity::Unary(), doc);
+  for (const auto& ty : NumericTypes()) {
+    auto output = is_integer(ty->id()) ? float64() : ty;
+    auto exec = IntToDoubleExecFromOp<ScalarUnaryNotNull, Op>(ty);
+    DCHECK_OK(func->AddKernel({ty}, output, exec));
+  }
+  return func;
+}
+
```

Comment:
I suggest to change the function name to `MakeUnaryArithmeticFunctionFloatOutTypeNotNull`.
Also, the `_checked` variants use the `ScalarUnaryNotNull` kernel exec generator but the regular variants use `ScalarUnary`. Need to add `MakeUnaryArithmeticFunctionFloatOutType` with same logic but using `ScalarUnary`.